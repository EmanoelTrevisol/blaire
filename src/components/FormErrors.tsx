import React from 'react';
import { StyleSheet, View, Text } from 'react-native';

export interface IFormErrorProps {
  messages: {
    required: string;
    valid: string;
  };
  error: boolean;
  errors: {
    valid: boolean;
    required: boolean;
  };
  dirty: boolean;
}

const FormErrors = (props: IFormErrorProps) => {
  const {
    error,
    dirty,
    messages,
    errors: { required, valid },
  } = props;
  if (!error || !dirty) {
    return null;
  }

  return (
    <View style={stl.errors}>
      {(required && (
        <Text style={stl.inputErrorText} testID="required-error">
          {messages.required}
        </Text>
      )) ||
        null}
      {(valid && (
        <Text style={stl.inputErrorText} testID="valid-error">
          {messages.valid}
        </Text>
      )) ||
        null}
    </View>
  );
};

const stl = StyleSheet.create({
  errors: {
    marginVertical: 10,
  },
  inputErrorText: {
    color: '#dd0000',
    fontSize: 12,
  },
});

export default FormErrors;
