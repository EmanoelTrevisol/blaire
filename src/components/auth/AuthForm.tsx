import React, { useState, useReducer, useEffect, useRef } from 'react';
import {
  StyleSheet,
  View,
  Text,
  TextInput,
  TouchableOpacity,
  ActivityIndicator,
} from 'react-native';

import { FormActionTypes, ActionTypes } from '@store/form/types';
import { formReducer } from '@store/form/reducer';
import initialFormState from '@store/form/state';

import FormErrors from '../form/FormErrors';

import { isValid } from '@validations/email';
import { matchesPattern } from '@validations/string';

const usernameErrorMessages = {
  required: 'Por favor, informe seu nome',
  valid: 'Ops... continue digitando. Estamos quase lá',
};

const emailErrorMessages = {
  required: 'Por favor, informe seu email',
  valid: 'Ops... esse email não parece ser válido',
};

const passwordErrorMessages = {
  required: 'Por favor, informe uma senha',
  valid: 'Quase lá! Falta pouco para cumprir os requisitos da senha',
};

function validateField({
  dispatcher,
  value,
  validateFn,
}: {
  dispatcher: React.Dispatch<FormActionTypes>;
  value: string;
  validateFn: Function;
}) {
  dispatcher({
    type: ActionTypes.SET_DIRTY,
    payload: {
      dirty: true,
    },
  });

  let error = false;
  const errors = {
    valid: false,
    required: false,
  };

  if (!value) {
    error = true;
    errors.required = true;
    errors.valid = false;
  } else if (!validateFn(value)) {
    error = true;
    errors.required = false;
    errors.valid = true;
  } else {
    error = false;
    errors.required = false;
    errors.valid = false;
  }

  dispatcher({
    type: ActionTypes.SET_ERRORS,
    payload: {
      error,
      errors,
    },
  });
}

export interface IAuthFormProps {
  isCreating?: boolean;
  onSubmit: Function;
  buttonTitle: string;
}

const LoginForm = (props: IAuthFormProps) => {
  const [isSubmitting, setIsSubmitting] = useState<boolean>(false);
  const [emailState, emailDispatcher] = useReducer(formReducer, {
    ...initialFormState,
  });
  const [usernameState, usernameDispatcher] = useReducer(formReducer, {
    ...initialFormState,
  });
  const [passwordState, passwordDispatcher] = useReducer(formReducer, {
    ...initialFormState,
  });

  const isMounted = useRef(true);

  useEffect(() => {
    return () => {
      isMounted.current = false;
    };
  }, []);

  const submit = async () => {
    validateUsername();
    validateEmail();
    validatePassword();
    if (emailState.error || passwordState.error) {
      return;
    }

    try {
      setIsSubmitting(true);
      await props.onSubmit({
        username: usernameState.value,
        email: emailState.value,
        password: passwordState.value,
      });
    } catch (error) {
      console.log('Error', error);
    } finally {
      if (isMounted.current) {
        setIsSubmitting(false);
      }
    }
  };

  const validateUsername = () => {
    validateField({
      dispatcher: usernameDispatcher,
      value: usernameState.value,
      validateFn: (username: string) => username.length > 3,
    });
  };

  const validateEmail = () => {
    validateField({
      dispatcher: emailDispatcher,
      value: emailState.value,
      validateFn: isValid,
    });
  };

  const validatePassword = () => {
    validateField({
      dispatcher: passwordDispatcher,
      value: passwordState.value,
      validateFn: (str: string) => {
        const patterns = [/[a-z]/, /[A-Z]/, /[0-9]/];
        return (
          str.length >= 8 &&
          patterns.every((pattern) => matchesPattern(str, pattern, 'g'))
        );
      },
    });
  };

  return (
    <>
      <View style={stl.form}>
        {(props.isCreating && (
          <View style={stl.formControl}>
            <View style={stl.label}>
              <Text style={stl.labelText}>Nome</Text>
            </View>
            <TextInput
              autoCapitalize="none"
              style={stl.input}
              value={usernameState.value}
              editable={!isSubmitting}
              keyboardType="default"
              textContentType="name"
              placeholder="Digite seu nome"
              placeholderTextColor="#898989"
              onBlur={validateUsername}
              onChangeText={(value) =>
                usernameDispatcher({
                  type: ActionTypes.SET_VALUE,
                  payload: { value },
                })
              }
            />
            <FormErrors messages={usernameErrorMessages} {...usernameState} />
          </View>
        )) ||
          null}
        <View style={stl.formControl}>
          <View style={stl.label}>
            <Text style={stl.labelText}>Email</Text>
          </View>
          <TextInput
            autoCapitalize="none"
            style={stl.input}
            value={emailState.value}
            editable={!isSubmitting}
            keyboardType="email-address"
            textContentType="username"
            placeholder="Digite seu email"
            placeholderTextColor="#898989"
            onBlur={validateEmail}
            onChangeText={(value) =>
              emailDispatcher({
                type: ActionTypes.SET_VALUE,
                payload: { value },
              })
            }
          />
          <FormErrors messages={emailErrorMessages} {...emailState} />
        </View>
        <View style={stl.formControl}>
          <View style={stl.label}>
            <Text style={stl.labelText}>Senha</Text>
            <Text style={stl.labelInstructions}>
              Sua senha deve conter pelo menos uma letra maiúscula e minúscula,
              um número e oito caracteres
            </Text>
          </View>
          <TextInput
            style={stl.input}
            value={passwordState.value}
            editable={!isSubmitting}
            textContentType={props.isCreating ? 'newPassword' : 'password'}
            placeholder="Insira sua senha"
            placeholderTextColor="#898989"
            passwordRules="required: lower; required: upper; required: digit; required: [-]; minlength: 6;"
            onChangeText={(value) =>
              passwordDispatcher({
                type: ActionTypes.SET_VALUE,
                payload: { value },
              })
            }
            secureTextEntry
            onBlur={validatePassword}
          />
          <FormErrors messages={passwordErrorMessages} {...passwordState} />
        </View>
      </View>
      <View style={stl.actions}>
        <TouchableOpacity
          style={stl.button}
          disabled={isSubmitting}
          onPress={submit}
        >
          {(isSubmitting && <ActivityIndicator color="#000" />) || (
            <Text style={stl.buttonText}>{props.buttonTitle}</Text>
          )}
        </TouchableOpacity>
      </View>
    </>
  );
};

const stl = StyleSheet.create({
  form: {
    marginBottom: 60,
  },
  formControl: {
    maxHeight: 80,
    marginVertical: 20,
  },
  label: {
    marginVertical: 5,
  },
  labelText: {
    fontSize: 20,
  },
  input: {
    paddingVertical: 5,
    borderBottomColor: '#000',
    borderBottomWidth: 1,
  },
  actions: {},
  button: {
    backgroundColor: '#ea94a5',
    padding: 16,
    borderRadius: 50,
    height: 50,
  },
  buttonText: {
    textAlign: 'center',
    fontWeight: 'bold',
  },
  labelInstructions: {
    fontSize: 12,
    color: '#232323',
  },
});

export default LoginForm;
