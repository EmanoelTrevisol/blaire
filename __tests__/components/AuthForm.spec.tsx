import React from 'react';
import AuthForm, { IAuthFormProps } from '@components/AuthForm';
import FormErrors from '@components/FormErrors';
import { act } from 'react-test-renderer';
import { fireEvent, render, waitFor } from '@testing-library/react-native';

const PASSWORD_INPUT_PLACEHOLDER = 'Insira sua senha';
const EMAIL_INPUT_PLACEHOLDER = 'Digite seu email';
let props: IAuthFormProps;

describe('AuthForm.tsx', () => {
  const renderComp = (cpProps: IAuthFormProps) => {
    return render(<AuthForm {...cpProps} />);
  };

  beforeEach(() => {
    props = {
      buttonTitle: 'Título do botão',
      onSubmit: jest.fn(),
    };
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  test('matches snapshot', () => {
    const { toJSON } = renderComp(props);

    expect(toJSON()).toMatchSnapshot();
  });

  describe('inputs and submition behavior', () => {
    describe('does not call onSubmit', () => {
      test('when email and password empty', async () => {
        const { getByText } = renderComp(props);

        const submitButton = getByText(props.buttonTitle);
        await waitFor(() => fireEvent.press(submitButton));

        expect(props.onSubmit).not.toHaveBeenCalled();
      });

      test('when only email is empty', async () => {
        const { queryByPlaceholderText, getByText } = renderComp(props);

        const submitButton = getByText(props.buttonTitle);
        const passwordInput = queryByPlaceholderText(
          PASSWORD_INPUT_PLACEHOLDER,
        )!;

        act(() => fireEvent.changeText(passwordInput, 'EssaEH4n0v4s3nh4'));
        act(() => fireEvent(passwordInput, 'blur'));

        await waitFor(() => fireEvent.press(submitButton));

        expect(props.onSubmit).not.toHaveBeenCalled();
      });

      test('when password is invalid', async () => {
        const { queryByPlaceholderText, getByText } = renderComp(props);

        const submitButton = getByText(props.buttonTitle);
        const passwordInput = queryByPlaceholderText(
          PASSWORD_INPUT_PLACEHOLDER,
        )!;

        act(() => fireEvent.changeText(passwordInput, 'abc'));
        act(() => fireEvent(passwordInput, 'blur'));

        await waitFor(() => fireEvent.press(submitButton));

        expect(props.onSubmit).not.toHaveBeenCalled();
      });

      test('when email is invalid', async () => {
        const { queryByPlaceholderText, getByText } = renderComp(props);

        const submitButton = getByText(props.buttonTitle);
        const emailInput = queryByPlaceholderText(EMAIL_INPUT_PLACEHOLDER)!;

        act(() => fireEvent.changeText(emailInput, 'email'));
        act(() => fireEvent(emailInput, 'blur'));

        await waitFor(() => fireEvent.press(submitButton));

        expect(props.onSubmit).not.toHaveBeenCalled();
      });
    });

    describe('calls onSubmit', () => {
      test('with valid email and password', async () => {
        props.onSubmit = jest.fn(() => Promise.resolve());

        const { queryByPlaceholderText, getByText } = renderComp(props);
        const args = {
          email: 'teste@test.com',
          password: 'EssaEH4n0v4s3nh4',
        };

        // const submitButton = ;
        const emailInput = queryByPlaceholderText(EMAIL_INPUT_PLACEHOLDER)!;
        const passwordInput = queryByPlaceholderText(
          PASSWORD_INPUT_PLACEHOLDER,
        )!;

        act(() => fireEvent.changeText(emailInput, args.email));
        act(() => fireEvent(emailInput, 'blur'));
        act(() => fireEvent.changeText(passwordInput, args.password));
        act(() => fireEvent(passwordInput, 'blur'));

        await act(() => fireEvent.press(getByText(props.buttonTitle)));

        expect(props.onSubmit).toHaveBeenCalledWith(args);
      });
    });

    describe('sends correct error props to FormErrors', () => {
      describe('email errors', () => {
        test('required', () => {
          const { getByPlaceholderText, UNSAFE_getAllByType } = renderComp(
            props,
          );

          const emailInput = getByPlaceholderText(EMAIL_INPUT_PLACEHOLDER);
          act(() => fireEvent(emailInput, 'blur'));

          const formErrors = UNSAFE_getAllByType(FormErrors);

          const formErrorProps = formErrors[0].props;

          expect(formErrorProps.dirty).toBe(true);
          expect(formErrorProps.error).toBe(true);
          expect(formErrorProps.errors.required).toBe(true);
          expect(formErrorProps.errors.valid).toBe(false);
        });
        test('valid', () => {
          const { getByPlaceholderText, UNSAFE_getAllByType } = renderComp(
            props,
          );

          const emailInput = getByPlaceholderText(EMAIL_INPUT_PLACEHOLDER);

          act(() => fireEvent.changeText(emailInput, 'email'));
          act(() => fireEvent(emailInput, 'blur'));

          const formErrors = UNSAFE_getAllByType(FormErrors);

          const formErrorProps = formErrors[0].props;

          expect(formErrorProps.dirty).toBe(true);
          expect(formErrorProps.error).toBe(true);
          expect(formErrorProps.errors.required).toBe(false);
          expect(formErrorProps.errors.valid).toBe(true);
        });
      });

      describe('password errors', () => {
        test('required', () => {
          const { getByPlaceholderText, UNSAFE_getAllByType } = renderComp(
            props,
          );

          const passwordInput = getByPlaceholderText(
            PASSWORD_INPUT_PLACEHOLDER,
          );

          act(() => fireEvent(passwordInput, 'blur'));

          const formErrors = UNSAFE_getAllByType(FormErrors);

          const formErrorProps = formErrors[1].props;

          expect(formErrorProps.dirty).toBe(true);
          expect(formErrorProps.error).toBe(true);
          expect(formErrorProps.errors.required).toBe(true);
          expect(formErrorProps.errors.valid).toBe(false);
        });

        test('valid', () => {
          const { getByPlaceholderText, UNSAFE_getAllByType } = renderComp(
            props,
          );

          const passwordInput = getByPlaceholderText(
            PASSWORD_INPUT_PLACEHOLDER,
          );

          act(() => fireEvent.changeText(passwordInput, 'email'));
          act(() => fireEvent(passwordInput, 'blur'));

          const formErrors = UNSAFE_getAllByType(FormErrors);

          const formErrorProps = formErrors[1].props;

          expect(formErrorProps.dirty).toBe(true);
          expect(formErrorProps.error).toBe(true);
          expect(formErrorProps.errors.required).toBe(false);
          expect(formErrorProps.errors.valid).toBe(true);
        });
      });

      describe('valid inputs', () => {
        test('valid email', () => {
          const { getByPlaceholderText, UNSAFE_getAllByType } = renderComp(
            props,
          );

          const emailInput = getByPlaceholderText(EMAIL_INPUT_PLACEHOLDER);

          act(() => fireEvent.changeText(emailInput, 'teste@teste.com'));
          act(() => fireEvent(emailInput, 'blur'));

          const formErrors = UNSAFE_getAllByType(FormErrors);

          const formErrorProps = formErrors[0].props;

          expect(formErrorProps.dirty).toBe(true);
          expect(formErrorProps.error).toBe(false);
          expect(formErrorProps.errors.required).toBe(false);
          expect(formErrorProps.errors.valid).toBe(false);
        });

        test('valid password', () => {
          const { getByPlaceholderText, UNSAFE_getAllByType } = renderComp(
            props,
          );

          const passwordInput = getByPlaceholderText(
            PASSWORD_INPUT_PLACEHOLDER,
          );

          act(() => fireEvent.changeText(passwordInput, 'Umas3nh4Valid4'));
          act(() => fireEvent(passwordInput, 'blur'));

          const formErrors = UNSAFE_getAllByType(FormErrors);

          const formErrorProps = formErrors[1].props;

          expect(formErrorProps.dirty).toBe(true);
          expect(formErrorProps.error).toBe(false);
          expect(formErrorProps.errors.required).toBe(false);
          expect(formErrorProps.errors.valid).toBe(false);
        });
      });
    });
  });
});
