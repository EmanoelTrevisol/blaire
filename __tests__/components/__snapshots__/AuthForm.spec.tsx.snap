// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthForm.tsx matches snapshot 1`] = `
Array [
  <View
    style={
      Object {
        "marginBottom": 60,
      }
    }
  >
    <View
      style={
        Object {
          "marginVertical": 20,
          "maxHeight": 80,
        }
      }
    >
      <View
        style={
          Object {
            "marginVertical": 5,
          }
        }
      >
        <Text
          style={
            Object {
              "fontSize": 20,
            }
          }
        >
          Email
        </Text>
      </View>
      <TextInput
        allowFontScaling={true}
        autoCapitalize="none"
        editable={true}
        keyboardType="email-address"
        onBlur={[Function]}
        onChangeText={[Function]}
        placeholder="Digite seu email"
        placeholderTextColor="#898989"
        rejectResponderTermination={true}
        style={
          Object {
            "borderBottomColor": "#000",
            "borderBottomWidth": 1,
            "paddingVertical": 5,
          }
        }
        textContentType="username"
        underlineColorAndroid="transparent"
        value=""
      />
    </View>
    <View
      style={
        Object {
          "marginVertical": 20,
          "maxHeight": 80,
        }
      }
    >
      <View
        style={
          Object {
            "marginVertical": 5,
          }
        }
      >
        <Text
          style={
            Object {
              "fontSize": 20,
            }
          }
        >
          Senha
        </Text>
        <Text
          style={
            Object {
              "color": "#232323",
              "fontSize": 12,
            }
          }
        >
          Sua senha deve conter pelo menos uma letra maiúscula e minúscula, um número e oito caracteres
        </Text>
      </View>
      <TextInput
        allowFontScaling={true}
        editable={true}
        onBlur={[Function]}
        onChangeText={[Function]}
        passwordRules="required: lower; required: upper; required: digit; required: [-]; minlength: 6;"
        placeholder="Insira sua senha"
        placeholderTextColor="#898989"
        rejectResponderTermination={true}
        secureTextEntry={true}
        style={
          Object {
            "borderBottomColor": "#000",
            "borderBottomWidth": 1,
            "paddingVertical": 5,
          }
        }
        textContentType="password"
        underlineColorAndroid="transparent"
        value=""
      />
    </View>
  </View>,
  <View
    style={Object {}}
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "#ea94a5",
          "borderRadius": 50,
          "height": 50,
          "opacity": 1,
          "padding": 16,
        }
      }
    >
      <Text
        style={
          Object {
            "fontWeight": "bold",
            "textAlign": "center",
          }
        }
      >
        Título do botão
      </Text>
    </View>
  </View>,
]
`;
